test/gurka/gurka.h

```diff
@@ -0,0 +1,632 @@
+            {osmium::item_type::node, node_osm_id_map[member.ref], member.role.c_str()});
+      } else {
+        if (way_osm_id_map.count(member.ref) == 0) {
+          throw std::runtime_error("Relation member refers to an undefined way " + member.ref);
+        }
+        members.push_back({osmium::item_type::way, way_osm_id_map[member.ref], member.role.c_str()});
+      }
+    }
+
+    std::vector<std::pair<std::string, std::string>> tags;
+    for (const auto& tag : relation.tags) {
+      tags.push_back({tag.first, tag.second});
+    }
+
+    osmium::builder::add_relation(buffer, osmium::builder::attr::_id(osm_id++),
+                                  osmium::builder::attr::_version(1),
+                                  osmium::builder::attr::_timestamp(std::time(nullptr)),
+                                  osmium::builder::attr::_members(members),
+                                  osmium::builder::attr::_tags(tags));
+  }
+
+  // Create header and set generator.
+  osmium::io::Header header;
+  header.set("generator", "valhalla-test-creator");
+
+  osmium::io::File output_file{filename, "pbf"};
+
+  // Initialize Writer using the header from above and tell it that it
+  // is allowed to overwrite a possibly existing file.
+  osmium::io::Writer writer{output_file, header, osmium::io::overwrite::allow};
+
+  // Write out the contents of the output buffer.
+  writer(std::move(buffer));
+
+  // Explicitly close the writer. Will throw an exception if there is
+  // a problem. If you wait for the destructor to close the writer, you
+  // will not notice the problem, because destructors must not throw.
+  writer.close();
+}
+
+} // namespace detail
+
+map buildtiles(const nodelayout layout,
+               const ways& ways,
+               const nodes& nodes,
+               const relations& relations,
+               const std::string& workdir) {
+
+  map result;
+  result.config = detail::build_config(workdir);
+  result.nodes = layout;
+
+  // Sanity check so that we don't blow away / by mistake
+  if (workdir == "/") {
+    throw std::runtime_error("Can't use / for tests, as we need to clean it out first");
+  }
+
+  if (boost::filesystem::exists(workdir))
+    boost::filesystem::remove_all(workdir);
+  boost::filesystem::create_directories(workdir);
+
+  auto pbf_filename = workdir + "/map.pbf";
+  std::cerr << "[          ] generating map PBF at " << pbf_filename << std::endl;
+  detail::build_pbf(result.nodes, ways, nodes, relations, pbf_filename);
+  std::cerr << "[          ] building tiles in " << result.config.get<std::string>("mjolnir.tile_dir")
+            << std::endl;
+  midgard::logging::Configure({{"type", ""}});
+
+  mjolnir::build_tile_set(result.config, {pbf_filename}, mjolnir::BuildStage::kInitialize,
+                          mjolnir::BuildStage::kValidate, false);
+
+  return result;
+}
+
+valhalla::Api
+route(const map& map, const std::vector<std::string>& waypoints, const std::string& costing) {
```

Comment:
It would be nice in the future if this could take a full valhalla::Options as input instead of just waypoints and costing. I'm trying to write a test locally that needs to set different `options.units` for kph and mph.

Reply:
I think the best way to implement this would be to allow the caller to supply a JSON template.  The call to `gurka::route()` would then inject the `locations` and `costings` property into that JSON.

I can add that as a follow-up PR to this - I'd like to get this whole thing merged first so I'm not constantly playing catch-up with the complex `test/astar.cc` test cases that keep changing with other recent work.